packages/core/src/cli/validate/upgradeability-assessment.ts

```diff
@@ -0,0 +1,138 @@
+import { getAnnotationArgs } from '../../utils/annotations';
+import { inferInitializable, inferUUPS } from '../../validate/query';
+import { ValidateCommandError } from './error';
+import { SourceContract } from './validations';
+
+export class ReferenceContractNotFound extends Error {
+  /**
+   * The contract reference that could not be found.
+   */
+  readonly reference: string;
+
+  /**
+   * The fully qualified name of the contract that referenced the missing contract.
+   */
+  readonly origin: string;
+
+  constructor(reference: string, origin: string) {
+    super(`Could not find contract ${reference} referenced in ${origin}.`);
+    this.reference = reference;
+    this.origin = origin;
+    this.name = 'ReferenceContractNotFound';
+  }
+}
+
+interface AnnotationAssessment {
+  upgradeable: boolean;
+  referenceName?: string;
+}
+
+export interface UpgradeabilityAssessment {
+  upgradeable: boolean;
+  referenceContract?: SourceContract;
+  uups?: boolean;
+}
+
+export function getUpgradeabilityAssessment(
+  contract: SourceContract,
+  allContracts: SourceContract[],
+): UpgradeabilityAssessment {
+  const fullContractName = contract.fullyQualifiedName;
+  const contractValidation = contract.validationData[fullContractName];
+
+  const isUUPS = inferUUPS(contract.validationData, fullContractName);
+
+  const annotationAssessment = getAnnotationAssessment(contract);
+  if (annotationAssessment.upgradeable) {
+    let referenceContract = undefined;
+    let isReferenceUUPS = false;
+    if (annotationAssessment.referenceName !== undefined) {
+      referenceContract = getReferenceContract(annotationAssessment.referenceName, contract, allContracts);
+      isReferenceUUPS = inferUUPS(referenceContract.validationData, referenceContract.fullyQualifiedName);
+    }
+
+    return {
+      upgradeable: true,
+      referenceContract: referenceContract,
+      uups: isReferenceUUPS || isUUPS,
+    };
+  } else {
+    const initializable = inferInitializable(contractValidation);
+    return {
+      upgradeable: initializable || isUUPS,
+      uups: isUUPS,
+    };
+  }
+}
+
+function getReferenceContract(reference: string, origin: SourceContract, allContracts: SourceContract[]) {
+  const referenceContracts = allContracts.filter(c => c.fullyQualifiedName === reference || c.name === reference);
```

Comment:
Does `allContracts` here consist of all contracts in the project/build or only the the upgradeable ones? I had some concerns about the linear search, but even if it's all contracts it's unlikely to be over 200 contracts so it's probably fine. Curious about your thoughts anyway.

Reply:
It consists of all contracts in the build, because it is looking for the reference which may or may not look like an upgradeable contract.  We could change this to something like a map.

Reply:
If the number is in the hundreds we are probably fine to keep as an array with linear search. Using a map would require a big refactor probably.